arch/arm/core/aarch32/cortex_a_r/exc_exit.S

```diff
@@ -93,10 +93,25 @@ SECTION_SUBSEC_FUNC(TEXT, _HandlerModeExit, z_arm_int_exit)
 	cmp r0, #1
 	bhi __EXIT_INT
 
+#ifdef CONFIG_USE_SWITCH
+	sub sp, sp, #0x4
+	mov r0, sp
+	bl z_arch_get_next_switch_handle
+	ldr r1, [sp]
+	add sp, sp, #0x4
+	cmp r0, r1
+	beq __EXIT_INT
+
+	push {r0-r1}
+	bl z_arm_pendsv
```

Comment:
I'm a little sad to see this still being implemented in terms of PendSV.  Are you guys aware of the CONFIG_SWAP_NONATOMIC issue?  The problem with using a low priority exception handler is that if you flag it and then release interrupts, you may **actually** be interrupted by a higher priority interrupt, which now has to run in a context where the kernel has set things up for a new _current thread expecting a context switch that didn't happen, and thus the thread that was interrupted is not actually what it thinks is _current!

There are a handful of workarounds in the scheduler for this condition, and almost every time I change something I manage to trip over this same edge case.  Would be nice to see it go away where possible. :)

Reply:
I hadn't heard of this issue before, but it sure sounds plausible that the SVC context can be interrupted by an IRQ. Maybe this could be avoided by using an IPI targeting the local CPU so that the separate SVC context isn't required any more? I'll give that a try, at least as a configurable alternative to what we have right now.

Reply:
It's not that the SVC can itself be interrupted (IIRC arm handlers start masked).  It's that it's priority is lower than hardware interrupts.  So if a timer (or whatever) fires while you have the scheduler lock and are deciding to context switch, and you call arch_switch() having made that decision, the next code to run is the timer interrupt, not the pendsv handler.

Now that timer interrupt, for example, decides to unblock a thread or whatever, so it needs to do some work checking the current/interrupted thread.  But that's now broken, because the "_current" variable for this CPU now reflects one thread (the one the scheduler decided to context switch to) but the thread that actually got interrupted is a **different** one (the thread from which the switch was going to happen).  And this causes bugs.  If you dig around in kernel/sched.c, you'll see a places where we have to stash off the "pending_current" thread pointer and then check underneath a kconfig to see if we hit this condition.  It's sort of a mess.

I tried to get folks to at least agree to use a highest-priority trap, but apparently that's difficult because someone (on Nordic, maybe?) had some non-zephyr handlers that needed to be able to set a PendSV in interrupt context and be guaranteed that the code wouldn't run until just before return to thread mode.

Really the right solution here is to abandon the exception handler technique and use a musical-chairs context switch like other architectures, IMHO.

But none of that is a crisis for this PR: as long as you have CONFIG_SWAP_NONATOMIC set, the scheduler will handle it correctly and the only cost is extra debugging for me on ARM every time I trip over it while making changes.

Reply:
I modified the arch/Kconfig once more so that SWAP_NONATOMIC is always selected if USE_SWITCH is selected. The use of USE_SWITCH is now no longer tied to the CPU, I've moved it to the board level so that it can be de-selected as both methods work on aarch32.